flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/Solver.scala

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.optimization
+
+import org.apache.flink.api.scala.DataSet
+import org.apache.flink.ml.common._
+import org.apache.flink.ml.math.{Vector => FlinkVector, BLAS, DenseVector}
+import org.apache.flink.api.scala._
+import org.apache.flink.ml.optimization.IterativeSolver._
+import org.apache.flink.ml.optimization.Solver._
+
+/** Base class for optimization algorithms
+ *
+ */
+abstract class Solver extends Serializable with WithParameters {
+
+  /** Provides a solution for the given optimization problem
+    *
+    * @param data A Dataset of LabeledVector (input, output) pairs
+    * @param initialWeights The initial weight that will be optimized
+    * @return A Vector of weights optimized to the given problem
+    */
+  def optimize(data: DataSet[LabeledVector], initialWeights: Option[DataSet[WeightVector]]):
+  DataSet[WeightVector]
+  // TODO(tvas): Maybe we want to pass a WeightVector directly here, instead of a
+  // DataSet[WeightVector]
```

Comment:
At the moment, I don't think that this is the case since the `collect` method is painfully slow.